### PR TITLE
fix(upd-cred): Allow `Update Credentials` to store in plain text if needed

### LINKED ISSUE #1750: Refresh required after Login operation
- [ ] ZE: After login, an extension refresh is required to be able to use the profile (since the config might have gotten updated)
- [ ] #1839
- [ ] #1838

_Originally posted by @zFernand0 in https://github.com/zowe/vscode-extension-for-zowe/issues/1743#issuecomment-1100440496_

### LINKED ISSUE #1804: Update profile does not work when non-secure credentials are missing in the first place
**Describe the bug**

When you disable secure credentials in a config file and then run the Update Credentials command it allows you to enter data and shows you a success message, but does not update the json file.

**To Reproduce**

- Create config file without secure fields and not autoload, no username and password properties
- Test that it works prompting you for credentials
- Right-click Update Credentials
- Enter credentials
- Success message
- Check the config file

It works when credentials are available in config file.

**Expected behavior**

Either an error message explaining that the properties are not available or the command adding the entered credentials to the profile. 

**Screenshots**

**Desktop (please complete the following information):**

- OS: Any
- Zowe Explorer Version: 2.0.1
- (Optional) Zowe CLI Version: 7.1.1
- (Optional) Are you using Secure Credential Store? no

**Additional context**

### LINKED ISSUE #1827: Prevent errors when `promptCredentials` operation is cancelled
**Describe the bug**

Whenever credentials are prompted, you will see multiple errors if you cancel out of the operation.
image

**To Reproduce**

Same as #1804 

**Expected behavior**

To only see the `Operation Cancelled` information message

**Desktop (please complete the following information):**

- OS: Any
- Zowe Explorer Version: 2.0.2
- (Optional) Zowe CLI Version: N/A
- (Optional) Are you using Secure Credential Store? No

**Additional context**

Encountered this issue while testing #1804

### LINKED ISSUE #1840: APIML token always stored in plain text after a login operation
**Describe the bug**

When you login to the APIML via ZE, the token always gets stored in plain text in the config file

**To Reproduce**

1. Login to the APIML
2. Check the Config file

**Expected behavior**

To store the token value securely if you are in a secure environment (where keytar can be used)

**Desktop (please complete the following information):**

- OS: Any
- Zowe Explorer Version: 2.0.3
- (Optional) Zowe CLI Version: N/A
- (Optional) Are you using Secure Credential Store? Yes

### PR DESCRIPTION
**Pull request by zFernand0**
## Proposed changes

Allow `Update Credentials` to store the user input in the team config file in an insecure environment
- Fix #1804
- Fix #1827
- Fix #1750
- Fix #1840

## Release Notes

Milestone: 2.0.4

Changelog: Fix `Update Credentials` by allowing it to store plain text secure fields.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [x] Any PR dependencies have been merged and published (if appropriate)
- [x] NEW: CLI version updated (again)

## GIFs

**Update Credentials**
temp

**Login and Logout**
temp

### CODE REVIEW

**Review by KutluOzel-b — DISMISSED** [review]
Tested each issue. LGTM :)

**Comment by JillieBeanSim** [comment]
@zFernand0 I see this in the theia test and may also refer to lint failure `ERROR: 1344:70  no-shadowed-variable  Shadowed name: 'prof'`

File: packages/zowe-explorer/src/Profiles.ts [hunk 1]
```diff
@@ -1327,26 +1331,20 @@ export class Profiles extends ProfilesCache {
...
-        const config = ImperativeConfig.instance.config;
-        const { user, global } = config.api.layers.find(upd.profileName);
-        const profilePath = config.api.profiles.expandPath(upd.profileName);
-        config.api.layers.activate(user, global);
-        config.set(`${profilePath}.properties.tokenType`, undefined);
-        config.set(`${profilePath}.properties.tokenValue`, undefined);
-        await config.save();
+        const mProfileInfo = await this.getProfileInfo();
+        const setSecure = mProfileInfo.isSecured();
+        const prof = mProfileInfo.getAllProfiles(profile.type).find((prof) => prof.profName === profile.name);
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/src/Profiles.ts:1344]: this is the only time I see `prof` in the changed code. in the const and as the find variable

Comment (zFernand0) [reply packages/zowe-explorer/src/Profiles.ts]: thanks! 😋 
Already push a quick fix for this!

**Comment by zFernand0** [comment]
In my latest commit I'm attempting to address the second piece of feedback about Update Credentials not working with `autoStore` false
Though I feel like it should work regardless of the `autoStore` value 🤔 
image

As for the first item...
I was considering creating the settings/imperative.json file, however...
1. When I edit the config to remove the plain text token and place the tokenValue in the secure array... 
2. And run the extRefresh as recommended by the info-message on creation of the Team Config... 
3. Then I can login just fine and the token is stored securely.

**Comment by JillieBeanSim** [comment]
> In my latest commit I'm attempting to address the second piece of feedback about Update Credentials not working with `autoStore` false Though I feel like it should work regardless of the `autoStore` value 🤔 image
> 
> As for the first item... I was considering creating the settings/imperative.json file, however...
> 
> 1. When I edit the config to remove the plain text token and place the tokenValue in the secure array...
> 2. And run the extRefresh as recommended by the info-message on creation of the Team Config...
> 3. Then I can login just fine and the token is stored securely.
> 
> **Note:** I left commented out code to create the `~/.zowe/settings/imperative.json` file (Secure by Default)

@zFernand0 I would think only write to the file if `autoStore:true` if `false` I would think it would work like before in v1 where we store prompted values in ProfilesCache.allProfiles for the VSC session.  So maybe it could be a way of updating credentials for the session only if `autoStore:false`, for instance if user changed pword on host while vsc was opened.  What do you think?

**Comment by zFernand0** [comment]
Discussion summary:
- The first issue was not reproduceable.
  - We would still want to have the means of creating all other missing folders (`~/.zowe/settings`)
- The second issue was mitigated by "disabling" the "Update Credentials" feature if `autoStore: false`
  - The idea of storing the credentials for the session alone seems to work for validation purposes (`checkCurrentProfile`)
    - However, some API calls may fail due to other TreeNode specific functionality that cannot be easily gathered (without much code duplication) from the `Update Credentials` command

**Review by JillieBeanSim — APPROVED** [review]
thanks for the updates and fixes @zFernand0

**Review by KutluOzel-b — APPROVED** [review]

**Pull request merged**